### PR TITLE
Port Segment Id/Name and SourceInfo.

### PR DESCRIPTION
**Pull request by scottweaver**
This set of changes aims to do the following.
- Bring over segment identifier and hashing logic from V 1.0
- Bring over source file information logic (previously PathData)
- Make code more idiomatic, specifically relocate struct construction logic to 'impl new' function.
- Add unit tests to all functions.
- Add rudimentary documentation.

### CODE REVIEW

**Review by PatrickJS-kunaico — APPROVED** [review]

**Review by TheWizardTower — COMMENTED** [review]
Had some questions (likely born of my own ignorance) but overall, I like the code you've written.

File: src/component.rs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+                } else if uscore {
+                    // Never push consecutive underscores.
+                    (acc, true)
+                } else {
+                    acc.push('_');
+                    (acc, true)
+                }
+            })
+            .0
+    }
```
[the remark this answers is not in the document]
Comment (scottweaver) [reply src/component.rs:48]: 1. The lossy bit comes straight over from the previous implementation so I assuming that this addressing a specific use-case.
2. I will look into a different name, it's not exposed so won't break anything.

File: src/component.rs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+                        normalized.push(component);
+                    }
+                }
+                _ => {
+                    normalized.push(component);
+                }
+            }
+        }
+        if ends_with_slash {
+            normalized.push("");
```
[the remark this answers is not in the document]
Comment (scottweaver) [reply src/component.rs:206]: Very good question.  Again, this is a direct port over so I assumed there was some weird edge case it might addressing which wasn't obvious.

Unfortunately, the original code base as no unit tests.   So, we can't look if there was a specific case it was addressing.

I tried to figure out in my added test cases if it had any real impact and I didn't notice anything when removing it.

I am going to say we can probably remove it.

File: src/component.rs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+        }
+        normalized
+    }
```
[the remark this answers is not in the document]
Comment (scottweaver) [reply src/component.rs]: Tested this and it fails if the files/dirs don't exist.  Not sure we want to introduced that as a requirement for this.

[the remark this answers is not in the document]
Comment (Adam-Kunai) [reply src/component.rs]: Completely fair.

**Review by Adam-Kunai — APPROVED** [review]
This looks solid. Excellent work.

**Review by wmertens — COMMENTED** [review]
Exciting :)

File: src/component.rs [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+/// Renamed from `PathData` in V 1.0.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SourceInfo {
+    pub abs_path: PathBuf,
+    pub rel_path: PathBuf,
+    pub abs_dir: PathBuf,
+    pub rel_dir: PathBuf,
+    pub file_stem: String,
+    pub extension: String,
+    pub file_name: String,
```
[the remark this answers is not in the document]
Comment (scottweaver) [reply src/component.rs:176]: From this I think I can remove the following:

- abs_path
- abd_dir
- file_stem
- file_extension
